pennylane/transforms/decompose.py

```diff
@@ -60,6 +60,77 @@ def _operator_decomposition_gen(
             )
 
 
+class DecomposeInterpreter(qml.capture.PlxprIntepreter):
+    """Plxpr Interpreter for applying the ``decompose`` transform to callables or jaxpr
+    when program capture is enabled.
+    """
+
+    def __init__(self, gate_set, max_expansion=None):
+        self.max_expansion = max_expansion
+
+        if isinstance(gate_set, (str, type)):
+            self.gate_set = set([gate_set])
+
+        if isinstance(gate_set, Iterable):
+            gate_types = tuple(gate for gate in gate_set if isinstance(gate, type))
+            gate_names = set(gate for gate in gate_set if isinstance(gate, str))
+            self.gate_set = lambda op: (op.name in gate_names) or isinstance(op, gate_types)
+
+        super().__init__()
+
+    def stopping_condition(self, op: qml.operation.Operator) -> bool:
+        """Function to determine whether or not an operator needs to be decomposed or not.
+
+        Args:
+            op (qml.operation.Operator): Operator to check.
+
+        Returns:
+            bool: Whether or not ``op`` is valid or needs to be decomposed. ``True`` means
+            that the operator does not need to be decomposed.
+        """
+        if not op.has_decomposition:
+            if not self.gate_set(op):
+                warnings.warn(
+                    f"Operator {op.name} does not define a decomposition and was not "
+                    f"found in the target gate set. To remove this warning, add the operator name "
+                    f"({op.name}) or type ({type(op)}) to the gate set.",
+                    UserWarning,
+                )
+            return True
+        return self.gate_set(op)
+
+    def interpret_operation(self, op: qml.operation.Operator):
+        """Interpret a PennyLane operation instance.
+
+        Args:
+            op (Operator): a pennylane operator instance
+
+        Returns:
+            Any
+
+        This method is only called when the operator's output is a dropped variable,
+        so the output will not affect later equations in the circuit.
+
+        See also: :meth:`~.interpret_operation_eqn`.
+
+        """
+        if self.gate_set(op):
+            return super().interpret_operation(op)
+
+        qml.capture.disable()
+        decomposition = list(
+            _operator_decomposition_gen(
+                op, self.stopping_condition, max_expansion=self.max_expansion
+            )
+        )
+        qml.capture.enable()
+
+        res = []
+        for decomp_op in decomposition:
+            res.append(super().interpret_operation(decomp_op))
+        return res
```

Comment:
I tried to use list comprehension for this like:
```python
return [super().interpret_operation(o) for o in decomposition]
```
but I kept getting the following error:
```
TypeError: super(type, obj): obj must be an instance or subtype of type
```

I fixed it by changing `super()` to `super(DecomposeInterpreter, self)`, but pylint raised an error about using python 3 syntax, so I just went with this.